Engine/Source/Settings.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 Genten Studios
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <Quartz2/Settings.hpp>
+#include <climits>
+
+using namespace q2;
+
+Setting::Setting(std::string name, std::string key, int defaultValue)
+    : m_name(std::move(name)), m_key(std::move(key)), m_value(defaultValue), m_maxValue(SHRT_MAX),
+      m_minValue(SHRT_MIN)
+{}
+
+bool Setting::set(int value)
+{
+	if (value >= m_minValue && value <= m_maxValue)
+	{
+		m_value = value;
+		return true;
+	}
+	return false;
+}
+
+void Setting::setMax(int value) { m_maxValue = value; }
+
+void Setting::setMin(int value) { m_minValue = value; }
+
+std::string Setting::getKey() { return m_key; }
+
+int Setting::value() { return m_value; }
+
+Setting* Settings::add(const std::string& name, const std::string& key, int defaultValue)
+{
+	m_settings.push_back(Setting(name, key, defaultValue));
+	return &m_settings.back();
+}
+
+const Setting* Settings::getSetting(const std::string& key)
+{
+	for (Setting& setting : m_settings)
```

Comment:
Since nothing in setting appears to be getting modified, this could be a const reference.
```suggestion
	for (const Setting& setting : m_settings)
```

Reply:
It invalidates the call to getKey(), we get this error on compiling:
`the object has type qualifiers that are not compatible with the member function "q2::Setting::getKey" -- object type is: const q2::Setting`

That being said, if we want to refactor this to use a map, we are no longer using this search method.

Reply:
Yeah that's because `getKey` is non const, if the method was made const as per a previous comment then you wouldn't get this error

Reply:
@DontBelieveMe I already fixed the getKey() but it's still not happy :(. Regardless I think I am going to attempt the map this morning.